libraries/common/io.effekt

```diff
@@ -63,96 +94,23 @@ namespace promise {
   extern io def make[T](): Promise[T] =
     js "promise$make()"
     llvm """
-      ; Get the default loop and run it.
       %promise = call %Pos @c_promise_make()
       ret %Pos %promise
     """
 
-  extern io def resolve[T](p: Promise[T], value: T): Unit =
-    js "${p}.resolve(${value})"
+  extern control def resolve[T](promise: Promise[T], value: T): Unit =
+    js "$effekt.callcc(callback => { ${promise}.resolve(${value}); callback($effekt.unit)})"
     llvm """
-      ; Get the default loop and run it.
-      call void @c_promise_resolve(%Pos %p, %Pos %value)
-      ret %Pos zeroinitializer
+      call void @c_promise_resolve(%Pos %promise, %Pos %value, %Stack %stack)
+      ret void
     """
 
-  extern io def await[T](p: Promise[T], callback: T => Unit at {io, global}): Unit =
-    js "${p}.promise.then(res => (${callback})(res).run())"
+  extern control def await[T](promise: Promise[T]): T =
+    js "$effekt.callcc(callback => ${promise}.promise.then(res => callback(res)))"
     llvm """
-      ; Get the default loop and run it.
-      call void @c_promise_await(%Pos %p, %Neg %callback)
-      ret %Pos zeroinitializer
+      call void @c_promise_await(%Pos %promise, %Stack %stack)
+      ret void
     """
 }
 
 
-// TODO we cannot add effects Files, Console, etc. to Task since modules cannot be recursive, at the moment.
-effect IO = { AsyncIO, Concurrent, Promises }
-
-type Task[T] = () => T / IO at {io, global}
-
-
-// Event Loop
-// ----------
-
-def eventloop(prog: Task[Unit]) = {
-
-  def handle(p: Task[Unit]): Unit =
-    try { p() }
-    with Concurrent {
-      def yield() = eventloop::schedule(box { resume(()) })
-      def abort() = ()
-    } with AsyncIO {
-      // TODO there is a bug in our LLVM backend where eta-reducing the following will break
-      def performIO[T](computation) = computation(box { t =>
-        eventloop::schedule(box { resume(t) })
-      })
-    } with Promises {
-      def promise[T](task: Task[T]) = {
-        val p = promise::make[T]();
-        handle(box {
-          val t = task()
-          promise::resolve(p, t)
-        })
-        eventloop::schedule( box { resume(p) })
-      }
-      def await[T](p: Promise[T]) = promise::await(p, box { t => resume(t) })
-    }
```

Comment:
One point of this exercise was to see how well we can write an eventloop within Effekt. I agree that for IO it could be built in. We just need to make sure (and test) that the semantics aligns between LLVM, JS, and what we intend.